src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/BitbucketDeploymentFactoryImpl.java

```diff
@@ -0,0 +1,76 @@
+package com.atlassian.bitbucket.jenkins.internal.deployments;
+
+import com.atlassian.bitbucket.jenkins.internal.model.deployment.BitbucketDeployment;
+import com.atlassian.bitbucket.jenkins.internal.model.deployment.BitbucketDeploymentEnvironment;
+import com.atlassian.bitbucket.jenkins.internal.model.deployment.DeploymentState;
+import com.google.common.annotations.VisibleForTesting;
+import hudson.model.Job;
+import hudson.model.Result;
+import hudson.model.Run;
+import jenkins.branch.MultiBranchProject;
+import org.jenkinsci.plugins.displayurlapi.DisplayURLProvider;
+
+import javax.annotation.CheckForNull;
+import java.util.Arrays;
+import java.util.Collection;
+
+public final class BitbucketDeploymentFactoryImpl implements BitbucketDeploymentFactory {
```

Comment:
Please remove `final` on all classes, I've only spotted it here and in `BitbucketDeploymentEnvironment` so far. I will not call it out again. final classes are a royal pain, you can't mock them, you can't extend them (which can be useful in testing or sometimes other places too). Sometimes that is desirable but those cases are far and few in between, and when they do come up we should add a comment at the top of the class explaining why it is final.